File: common/src/main/java/dev/mayaqq/estrogen/client/features/boobs/BoobPhysicsManager.java
package dev.mayaqq.estrogen.client.features.boobs;

import dev.mayaqq.estrogen.config.EstrogenConfig;
import dev.mayaqq.estrogen.utils.Boob;
import net.minecraft.client.Minecraft;
import net.minecraft.client.multiplayer.ClientLevel;
import net.minecraft.world.entity.player.Player;

import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;

public class BoobPhysicsManager {

    private static final ConcurrentHashMap<UUID, Physics> players = new ConcurrentHashMap<>();

    public static boolean isEnabled() {
        return EstrogenConfig.client().chestPhysicsRendering.get();
    }

    public static void tick() {
        ClientLevel level = Minecraft.getInstance().level;
        if (level == null) return;
        if (!isEnabled()) return;
        for (ConcurrentHashMap.Entry<UUID, Physics> physics : players.entrySet()) {
            Player player = level.getPlayerByUUID(physics.getKey());
            if (player != null && Boob.shouldShow(player)) {
                physics.getValue().update(player);
                if (physics.getValue().expired) {
                    players.remove(physics.getKey());
                }
            } else {
                players.remove(physics.getKey());
            }
        }
    }

    public static Physics getPhysicsForPlayer(Player player) {
        return players.computeIfAbsent(player.getUUID(), uuid -> new Physics());
    }
}

File: src/main/kotlin/dev/mayaqq/estrogen/client/features/boobs/BoobPhysicsManager.kt
package dev.mayaqq.estrogen.client.features.boobs

import dev.mayaqq.cynosure.client.events.ClientTickEvent
import dev.mayaqq.cynosure.events.api.Subscription
import dev.mayaqq.cynosure.helpers.McLevel
import dev.mayaqq.estrogen.config.EstrogenClientConfig
import net.minecraft.world.entity.player.Player
import java.util.*
import java.util.concurrent.ConcurrentHashMap


object BoobPhysicsManager {
    private val players = ConcurrentHashMap<UUID, Physics>()

    fun isEnabled(): Boolean = EstrogenClientConfig.ChestRenderingGlobal.physicsRendering

    @Subscription
    fun tick(event: ClientTickEvent.Begin) {
        val level = McLevel?: return
        if (!isEnabled()) return

        for (physics in players.entries) {
            val player = level.getPlayerByUUID(physics.key)
            if (player != null && Boob.shouldShow(player)) {
                physics.value.update(player)
                if (physics.value.expired) {
                    players.remove(physics.key)
                }
            } else {
                players.remove(physics.key)
            }
        }
    }

    fun getPhysicsForPlayer(player: Player): Physics {
        return players.computeIfAbsent(player.uuid) { _ -> Physics() }
    }
}